codex/blockexchange/engine/advertiser.nim

```diff
@@ -129,9 +137,11 @@ proc start*(b: Advertiser) {.async.} =
 
   b.advertiserRunning = true
   for i in 0..<b.concurrentAdvReqs:
-    b.advertiseTasks.add(processQueueLoop(b))
+    let fut = b.processQueueLoop().track(b)
+    asyncSpawn fut
```

Comment:
I confess I'm always weirded out when I see this. My simple brain parses that the Future returned by `processQueuedLoop` is tracking something, when in fact there's this unexpected inversion and it's actually b.trackedFutures that is tracking b.processQueueLoop.

Reply:
Sure, the API could use some work, but that can be handled in another PR and applied across the codebase.

Reply:
There are reasons the API ended up where we are now, but those reasons are being removed from the codebase, so we'll have some freedom to change the API (not in this PR).

Something like this would be better?
```nim
let fut = b.processQueueLoop()
b.trackedFutures.track(fut)
asyncSpawn fut
```

Reply:
Looking just at the code:

```nim
let fut = b.processQueueLoop().track(b)
```

it is not evident what it does. To understand you have to check the return type of `processQueueLoop` which apparently has ability to “track” objects of the type corresponding to `b`. The reader really needs to dissect the code to understand.

The proposed alternative is way easier to understand. You could even just say `b.trackedFutures.add(fut)` or maybe even instead of `trackedFutures` something have a type `FutureTracker` and then `b.futureTracker.add(fut)` or maybe there could be a special type of a “TrackedFuture” which would take care of itself magically (I do not know if that would work here - I did not dive into details)…

Reply:
>  maybe there could be a special type of a “TrackedFuture” which would take care of itself magically

There are definitely ways to do this, however I'm not a fan of magic things in most cases because they tend to hide what's happening and make things harder to reason, harder to debug, and harder to add new features (even though it's usually intended to make things easier to implement, however it's rarely the case).

Reply:
When saying "magic" I mean more something in the style of RAII (Resource Acquisition is Initialization) common in C++ and less in the style of Java/C#/Spring like fancy annotations.

Reply:
PR here: https://github.com/codex-storage/nim-codex/pull/1046